core/src/splitting/CausalSurvivalSplittingRule.cpp

```diff
@@ -0,0 +1,322 @@
+/*-------------------------------------------------------------------------------
+  This file is part of generalized random forest (grf).
+
+  grf is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  grf is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with grf. If not, see <http://www.gnu.org/licenses/>.
+ #-------------------------------------------------------------------------------*/
+
+#include <algorithm>
+#include <cmath>
+
+#include "CausalSurvivalSplittingRule.h"
+
+namespace grf {
+
+CausalSurvivalSplittingRule::CausalSurvivalSplittingRule(size_t max_num_unique_values,
+                                                         uint min_node_size,
+                                                         double alpha,
+                                                         double imbalance_penalty):
+    min_node_size(min_node_size),
+    alpha(alpha),
+    imbalance_penalty(imbalance_penalty) {
+  this->counter = new size_t[max_num_unique_values];
+  this->weight_sums = new double[max_num_unique_values];
+  this->sums = new double[max_num_unique_values];
+  this->num_small_z = new size_t[max_num_unique_values];
+  this->sums_z = new double[max_num_unique_values];
+  this->sums_z_squared = new double[max_num_unique_values];
+  this->failure_count = new size_t[max_num_unique_values];
+}
+
+CausalSurvivalSplittingRule::~CausalSurvivalSplittingRule() {
+  if (counter != nullptr) {
+    delete[] counter;
+  }
+  if (weight_sums != nullptr) {
+    delete[] weight_sums;
+  }
+  if (sums != nullptr) {
+    delete[] sums;
+  }
+  if (sums_z != nullptr) {
+    delete[] sums_z;
+  }
+  if (sums_z_squared != nullptr) {
+    delete[] sums_z_squared;
+  }
+  if (num_small_z != nullptr) {
+    delete[] num_small_z;
+  }
+  if (failure_count != nullptr) {
+    delete[] failure_count;
+  }
+}
+
+bool CausalSurvivalSplittingRule::find_best_split(const Data& data,
+                                                  size_t node,
+                                                  const std::vector<size_t>& possible_split_vars,
+                                                  const std::vector<double>& responses_by_sample,
+                                                  const std::vector<std::vector<size_t>>& samples,
+                                                  std::vector<size_t>& split_vars,
+                                                  std::vector<double>& split_values,
+                                                  std::vector<bool>& send_missing_left) {
```

Comment:
The other splitting rules have tests in `SplitRuleInvarianceTest` for missing values -- we could add a case for this new rule.

Reply:
Yes, I've added one now!